### PR TITLE
Switch Nearby to Wikidata Query

### LINKED ISSUE #289: Switch to Wikidata in addition instead of wiki-needs-pictures
I started to wonder whether we should use Wikidata instead of wiki-needs-pictures.

Advantages of Wikidata:
- Real-time. Add a picture to the item, and it disappears from the map immediately. wiki-needs-pictures is only updated once a week.
- Fast: In my tests the WikiData request takes about 2 seconds (even in an area with many items needing pictures), compared with wiki-needs-pictures' 17 seconds on the same network connection.
- Data plan-friendly: Wikidata uses only a few kilobytes, whereas wiki-needs-pictures downloads a file of several megabytes every time the app is started. This is thanks to WikiData has out-of-the-box area queries. Implementing area queries for wiki-needs-pictures would require a lot of development and a good server.
- Memory-friendly: Not having to keep this multimegabytes file in memory.
- Works for all Wikipedias, unlike wiki-needs-pictures which needs custom code for each Wikipedia (I added that code for the Japanese Wikipedia)

Advantages of wiki-needs-pictures:
- Listed pictures really need someone to go there and take a picture (whereas on Wikidata most missing pictures are actually available on Commons and just not linked yet, see [WDFIST](https://tools.wmflabs.org/fist/wdfist/?wdq=AROUND[625,35.66262,139.73060,2]&depth=3&language=en&project=wikipedia&no_images_only=1&prefilled=1)).
- Special requests may be implemented in the future. For instance: there are are already pictures of this temple's exterior, but a picture of the inside is needed.

---

All of the API requests we need (and more) can be found in the new WikiShootMe: https://tools.wmflabs.org/wikishootme/

When [this WDFIST issue](https://meta.wikimedia.org/wiki/Talk:WDFIST#E1:_Bad_API_response_mwoauth-invalid-authorization-invalid-user_despite_WiDaR_authorized.2C_on_Android) is fixed, I think we should also link to WDFIST like WikiShootMe does. For Wikidata items that do not have an image, it proposes some found on the Wikipedia articles linked to this item. Extremely useful before deciding whether walking that 2 kilometers detour is really worth it.

### LINKED ISSUE #485: Pre-release Nearby does not detect location
The pre-release version of Nearby (current master branch) loads the CSV, but does not detect your location when used in a freshly-installed app. Tested on an API 24 real device (Samsung Galaxy S7). The released version works fine (tested on the same night).

Might have something to do with location permissions not being asked? Not sure, will have to test further. @maskaravivek , what do you think could have broken it?

### PR DESCRIPTION
**Pull request by whym**
Fixes #289 partially - not a full migration but a first step. Wikidata-based results seem to be less accurate - it sometimes suggests articles with images on Wikipedia pages (but without an image *on Wikidata*). I think further filtering is necessary.

This may fix #485 too. (It works for me, but please test.)

### CODE REVIEW

File: app/src/main/java/fr/free/nrw/commons/nearby/NearbyPlaces.java [hunk 1]
```diff
@@ -9,23 +10,156 @@
...
 public class NearbyPlaces {
 
     private static final String TAG = NearbyPlaces.class.getName();
-    static List<Place> places = null;
+    private static final int MIN_RESULTS = 20;
+    private static final int MAX_RESULTS = 500;
+    private static final double RADIUS_INITIAL = 0.5;
+    private static final double RADIUS_MULTIPLIER = 3.0;
+    private static final String WIKIDATA_QUERY_URL = "https://query.wikidata.org/sparql?query={{QUERY}}";
+    private static final String WIKIDATA_QUERY_TEMPLATE =
```
Comment (whym) [inline app/src/main/java/fr/free/nrw/commons/nearby/NearbyPlaces.java]: I realize this is ugly (and produces a lot of code style warnings) but couldn't think of a quick alternative.

**Comment by misaochan** [comment]
The `getFromWikiNeedsPictures()` option works well for me and solves the problem. :) The `getFromWikidataQuery()` option crashes with the log below (I guess this is why you set the preference by default to use `getFromWikiNeedsPictures()`)? Regardless, I'm good with merging if @nicolas-raoul is.

```
04-14 00:46:13.717: E/AndroidRuntime(26643): FATAL EXCEPTION: AsyncTask #2
04-14 00:46:13.717: E/AndroidRuntime(26643): Process: fr.free.nrw.commons, PID: 26643
04-14 00:46:13.717: E/AndroidRuntime(26643): java.lang.RuntimeException: An error occurred while executing doInBackground()
04-14 00:46:13.717: E/AndroidRuntime(26643): 	at android.os.AsyncTask$3.done(AsyncTask.java:318)
04-14 00:46:13.717: E/AndroidRuntime(26643): 	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
04-14 00:46:13.717: E/AndroidRuntime(26643): 	at java.util.concurrent.FutureTask.setException(FutureTask.java:223)
04-14 00:46:13.717: E/AndroidRuntime(26643): 	at java.util.concurrent.FutureTask.run(FutureTask.java:242)
04-14 00:46:13.717: E/AndroidRuntime(26643): 	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:243)
04-14 00:46:13.717: E/AndroidRuntime(26643): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
04-14 00:46:13.717: E/AndroidRuntime(26643): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
04-14 00:46:13.717: E/AndroidRuntime(26643): 	at java.lang.Thread.run(Thread.java:762)
04-14 00:46:13.717: E/AndroidRuntime(26643): Caused by: java.lang.RuntimeException: java.io.FileNotFoundException: https://query.wikidata.org/sparql?query=SELECT%0A++%28SAMPLE%28%3Flocation%29+as+%3Flocation%29%0A++%3Fitem%0A++%28SAMPLE%28COALESCE%28%3Fitem_label_preferred_language%2C+%3Fitem_label_any_language%29%29+as+%3Flabel%29%0A++%28SAMPLE%28%3FclassId%29+as+%3Fclass%29%0A++%28SAMPLE%28COALESCE%28%3Fclass_label_preferred_language%2C+%3Fclass_label_any_language%2C+%22%3F%22%29%29+as+%3Fclass_label%29%0A++%28SAMPLE%28COALESCE%28%3Ficon0%2C+%3Ficon1%29%29+as+%3Ficon%29%0A++%28SAMPLE%28COALESCE%28%3Femoji0%2C+%3Femoji1%29%29+as+%3Femoji%29%0AWHERE+%7B%0A++%23+Around+given+location...%0A++SERVICE+wikibase%3Aaround+%7B%0A++++%3Fitem+wdt%3AP625+%3Flocation.%0A++++bd%3AserviceParam+wikibase%3Acenter+%22Point%28153.0403590974966+-27.453153960089494%29%22%5E%5Egeo%3AwktLiteral.+%0A++++bd%3AserviceParam+wikibase%3Aradius+%221093.5%22+.+%23+Radius+in+kilometers.%0A++%7D%0A++%0A++%23+...+and+without+an+image.%0A++MINUS+%7B%3Fitem+wdt%3AP18+%5B%5D%7D%0A++%0A++%23+Get+the+label+in+the+preferred+language+of+the+user%2C+or+any+other+language+if+no+label+is+available+in+that+language.%0A++OPTIONAL+%7B%3Fitem+rdfs%3Alabel+%3Fitem_label_preferred_language.+FILTER+%28lang%28%3Fitem_label_preferred_language%29+%3D+%22en%22%29%7D%0A++OPTIONAL+%7B%3Fitem+rdfs%3Alabel+%3Fitem_label_any_language%7D%0A++%0A++%23+Get+the+class+label+in+the+preferred+language+of+the+user%2C+or+any+other+language+if+no+label+is+available+in+that+language.%0A++OPTIONAL+%7B%0A++++%3Fitem+p%3AP31%2Fps%3AP31+%3FclassId.%0A++++OPTIONAL+%7B%3FclassId+rdfs%3Alabel+%3Fclass_label_preferred_language.+FILTER+%28lang%28%3Fclass_label_preferred_language%29+%3D+%22en%22%29%7D%0A++++OPTIONAL+%7B%3FclassId+rdfs%3Alabel+%3Fclass_label_any_language%7D%0A%0A++++%23+Get+icon%0A++++OPTIONAL+%7B+%3FclassId+wdt%3AP2910+%3Ficon0.+%7D%0A++++OPTIONAL+%7B+%3FclassId+wdt%3AP279*%2Fwdt%3AP2910+%3Ficon1.+%7D%0A++++%23+Get+emoji%0A++++OPTIONAL+%7B+%3FclassId+wdt%3AP487+%3Femoji0.+%7D%0A++++OPTIONAL+%7B+%3FclassId+wdt%3AP279*%2Fwdt%3AP487+%3Femoji1.+%7D%0A++%7D%0A%7D%0AGROUP+BY+%3Fitem%0A
04-14 00:46:13.717: E/AndroidRuntime(26643): 	at fr.free.nrw.commons.nearby.NearbyPlaces.getFromWikidataQuery(NearbyPlaces.java:149)
04-14 00:46:13.717: E/AndroidRuntime(26643): 	at fr.free.nrw.commons.nearby.NearbyPlaces.getFromWikidataQuery(NearbyPlaces.java:79)
04-14 00:46:13.717: E/AndroidRuntime(26643): 	at fr.free.nrw.commons.nearby.NearbyListFragment.loadAttractionsFromLocation(NearbyListFragment.java:199)
04-14 00:46:13.717: E/AndroidRuntime(26643): 	at fr.free.nrw.commons.nearby.NearbyListFragment.access$200(NearbyListFragment.java:32)
04-14 00:46:13.717: E/AndroidRuntime(26643): 	at fr.free.nrw.commons.nearby.NearbyListFragment$NearbyAsyncTask.doInBackground(NearbyListFragment.java:148)
04-14 00:46:13.717: E/AndroidRuntime(26643): 	at fr.free.nrw.commons.nearby.NearbyListFragment$NearbyAsyncTask.doInBackground(NearbyListFragment.java:126)
04-14 00:46:13.717: E/AndroidRuntime(26643): 	at android.os.AsyncTask$2.call(AsyncTask.java:304)
04-14 00:46:13.717: E/AndroidRuntime(26643): 	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
04-14 00:46:13.717: E/AndroidRuntime(26643): 	... 4 more
```

**Comment by whym** [comment]
That's strange, the Wikdata-based one should work. I'm investigating it.

File: app/src/main/java/fr/free/nrw/commons/nearby/NearbyPlaces.java [hunk 2]
```diff
@@ -9,23 +10,156 @@
...
+        List<Place> places = Collections.emptyList();
+        while (true) {
+            places = getFromWikidataQuery(curLatLng, lang, radius);
+            if (places.size() >= MIN_RESULTS) {
+                break;
+            } else {
+                radius *= RADIUS_MULTIPLIER;
+            }
+        }
+        while (places.size() > MAX_RESULTS && radius > RADIUS_INITIAL) {
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/nearby/NearbyPlaces.java]: Is there any drawback to getting "too many" places? If the places are sorted from nearest (or on a map) there should be no problem I guess?

File: app/src/main/java/fr/free/nrw/commons/nearby/NearbyPlaces.java [hunk 3]
```diff
@@ -9,23 +10,156 @@
...
+
+    private static List<Place> getFromWikidataQuery(LatLng curLatLng, String lang, double radius) {
+        try {
+            String query = WIKIDATA_QUERY_TEMPLATE.replace("{{RADIUS}}", "" + radius)
+                    .replace("{{LAT}}", "" + curLatLng.latitude)
+                    .replace("{{LONG}}", "" + curLatLng.longitude)
+                    .replace("{{LANG}}", "" + lang);
+            query = URLEncoder.encode(query, "utf-8");
+            String url = WIKIDATA_QUERY_URL.replace("{{QUERY}}", query);
+            Log.v(TAG, url.toString());
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/nearby/NearbyPlaces.java]: Would you mind logging the radius too? I guess fine-tuning the radius one of the most important reasons to read the log, and right now it is very hard to read from that long query string. Thanks a lot!

**Comment by nicolas-raoul** [comment]
I believe the radius got too large: The query found in @misaochan's log times out when I try it.
3.0 is a very extreme radius multiplier, as the number of results statistically grows with the square of the radius. How about 1.5 instead?

File: app/src/main/java/fr/free/nrw/commons/nearby/NearbyPlaces.java (hunk 2 above)
Comment (whym) [inline app/src/main/java/fr/free/nrw/commons/nearby/NearbyPlaces.java]: This is mainly a prevention against errors on the server side. For too many results (e.g. tens of thousands), sometimes the host may terminate it, giving no result to us.

**Comment by whym** [comment]
I found that the reason not working for @misaochan was that I didn't take into account negative latitude values. (I blame my own northern hemisphere point of view. ;))

Nothing was found while the radius kept increasing.

**Comment by misaochan** [comment]
Ah okay, that makes sense haha. BTW, when this is fixed, how do we want to handle the results from both methods? Should we default to Wikidata and give the user an option to switch to WNP (via an icon on the action bar maybe?).

File: app/src/main/java/fr/free/nrw/commons/nearby/NearbyPlaces.java (hunk 3 above)
Comment (whym) [inline app/src/main/java/fr/free/nrw/commons/nearby/NearbyPlaces.java]: Good point, will do.

**Comment by whym** [comment]
I don't have a good suggestion about it yet. The option to choose Wikidata/WNP was meant to be a temporary one - to allow people to compare the two results and hopefully come up with an idea to combine the two (and maybe other tools like the PageImages API).

**Comment by nicolas-raoul** [comment]
I would strongly suggest enabling Wikidata and disabling WNP, by default. Because it uses less data.
Our users in Bangladesh/etc will be surprised to get all of their monthly mobile data eaten by this feature.
I recently went to Morocco (which is a very advanced country) but even there most people use mobile data, with small data plans. Even in city centers people having access to landline-backed WiFi are a minority. Most households have a WiFi router on which they plug a mobile dongle.

**Comment by misaochan** [comment]
Yeah, let me know when the bug is fixed and I'll test it again. :)

For handling the results, maybe we could keep the current temporary fix of having the option to choose in Settings, but make Wikidata the default instead? Then in Settings we can explain a bit to users and warn them that choosing to use WNP might drain their mobile data. Most users will probably not bother to try WNP in that case, but those who want to help us test it might be able to, which could be helpful for the initial few months.

**Comment by whym** [comment]
Thanks, I'm convinced. Actually the latest commit here already incorporates the suggestion of making Wikidata default. The crash should have been fixed too.  Please try again!

**Comment by misaochan** [comment]
Works brilliantly now, thanks @whym ! Will release to beta ASAP.

**Pull request merged**